### PR TITLE
[AMORO-1035][Flink] Support customizing mixed-format table source parallelism

### LINKED ISSUE #1035: [Improvement][Flink]: Support customizing Arctic source parallelism
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Support customizing Arctic source parallelism, especially for log source.

### How should we improve?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by YesOrNo828**
## Why are the changes needed?

Support customizing the parallelism of the filestore/logstore source via Flink SQL.

Close #1035 .

## Brief change log

- update flink modules,
- update the filestore and logstore table source

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no) yes
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) docs

### CODE REVIEW

**Review by lklhdu — APPROVED** [review]
LGTM.

**Comment by huyuanfeng2018** [comment]
Can you show the rendering of flink webui after the change? As far as I know, Flinksql has not yet supported custom source parallelism. In flinksql, all operators use the parallelism of the previous operator as the current parallelism, so After I customized the source.parallelism , I think the parallelism of subsequent tasks will not run according to the default parallelism of the task.

**Comment by YesOrNo828** [comment]
> Can you show the rendering of flink webui after the change? As far as I know, Flinksql has not yet supported custom source parallelism. In flinksql, all operators use the parallelism of the previous operator as the current parallelism, so After I customized the source.parallelism , I think the parallelism of subsequent tasks will not run according to the default parallelism of the task.

@huyuanfeng2018  Thanks for your comment. Actually we could modify the source operater's parallelism through the Table API instead of Flink SQL.  In the `ScanTableSource.getScanRuntimeProvider()` method, we could modify the DataStreamSource's parallelism.

image

```sql
select * from 
log_data /*+OPTIONS('arctic.read.mode'='log','properties.group.id'='......','parallelism'='5' )*/ --The job parallelism is 10.
```
Specify source parallelism with SQL hint

**Comment by huyuanfeng2018** [comment]
> > Can you show the rendering of flink webui after the change? As far as I know, Flinksql has not yet supported custom source parallelism. In flinksql, all operators use the parallelism of the previous operator as the current parallelism, so After I customized the source.parallelism , I think the parallelism of subsequent tasks will not run according to the default parallelism of the task.
> 
> @huyuanfeng2018 Thanks for your comment. Actually we could modify the source operater's parallelism through the Table API. In the `ScanTableSource.getScanRuntimeProvider()` method, we could modify the DataStreamSource's parallelism.
> 
> image
> ```sql
> select * from 
> log_data /*+OPTIONS('arctic.read.mode'='log','properties.group.id'='......','parallelism'='5' )*/ --The job parallelism is 10.
> ```
> 
> Specify source parallelism with SQL hint

What I am confused about is how the subsequent task changed back to 10 after the scan parallelism was changed? Let me give you an example. From the perspective of flink-table-planner, often a `select` statement in the source table will trigger `Calc`. However, the parallelism in `Calc` is the parallelism of the source Transformation used.:
image
I didn’t see this modification in this PR. I would like to know how this is done. In our internal practice, we need to make certain modifications to Flink-table-planner to achieve this effect.

**Comment by YesOrNo828** [comment]
> > > Can you show the rendering of flink webui after the change? As far as I know, Flinksql has not yet supported custom source parallelism. In flinksql, all operators use the parallelism of the previous operator as the current parallelism, so After I customized the source.parallelism , I think the parallelism of subsequent tasks will not run according to the default parallelism of the task.
> > 
> > 
> > @huyuanfeng2018 Thanks for your comment. Actually we could modify the source operater's parallelism through the Table API. In the `ScanTableSource.getScanRuntimeProvider()` method, we could modify the DataStreamSource's parallelism.
> > image
> > ```sql
> > select * from 
> > log_data /*+OPTIONS('arctic.read.mode'='log','properties.group.id'='......','parallelism'='5' )*/ --The job parallelism is 10.
> > ```
> > Specify source parallelism with SQL hint
> 
> What I am confused about is how the subsequent task changed back to 10 after the scan parallelism was changed? Let me give you an example. From the perspective of flink-table-planner, often a `select` statement in the source table will trigger `Calc`. However, the parallelism in `Calc` is the parallelism of the source Transformation used.: image I didn’t see this modification in this PR. I would like to know how this is done. In our internal practice, we need to make certain modifications to Flink-table-planner to achieve this effect.

Because the above example has a window aggregate operator, it can't be chained with the previous source and calc operators, so it will become 10 parallelism, and the source and calc operators are 5 parallelism. This PR does not change the parallelism of the source operator alone, but also the operators in its chain.

**Review by huyuanfeng2018 — APPROVED** [review]
LGTM.

**Comment by YesOrNo828** [comment]
Fixed the conflicts after the spotless tool was added to the Flink module.

File: mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/table/descriptors/ArcticValidator.java [hunk 1]
```diff
@@ -289,6 +289,15 @@ public class ArcticValidator extends ConnectorDescriptorValidator {
                   TableFormat.PAIMON,
                   TableFormat.MIXED_ICEBERG));
 
+  public static final ConfigOption<Integer> SCAN_PARALLELISM =
+      ConfigOptions.key("parallelism")
```
Comment (czy006) [inline mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/table/descriptors/ArcticValidator.java:293]: maybe `source.parallelism` name better than `parallelism` ?

**Review by czy006 — CHANGES_REQUESTED** [review]

**Review by czy006 — APPROVED** [review]

**Comment by czy006** [comment]
LGTM

**Review by zhoujinsong — APPROVED** [review]
LGTM.
